bucketers.py

```diff
@@ -156,6 +156,37 @@ def calc_source_bucketed_matches(self, src, ref, out, ref_aligns, out_aligns, sr
         fmeas = 2 * prec * rec / (prec + rec)
       yield both_tot, ref_tot, out_tot, rec, prec, fmeas
 
+  def calc_bucketed_likelihoods(self, corpus, likelihoods):
+    """
+    Calculate the average of log likelihoods, bucketed by the type of word/label we have
+    This must be used with a subclass that has self.bucket_strs defined, and self.calc_bucket(word) implemented.
+
+    Args:
+      corpus: The text/label corpus over which we compute the likelihoods
+      likelihoods: The log-likelihoods corresponding to each word/label in the corpus
+
+    Returns:
+      the average log-likelihood bucketed by the type of word/label we have
+    """
+
+    if type(corpus) == str:
+      corpus = corpus_utils.load_tokens(corpus)
+    bucketed_likelihoods = [[0.0, 0] for _ in self.bucket_strs]
+    assert (len(corpus) == len(likelihoods))
```

Comment:
As you may find in other parts of the codebase, this project prefers `raise ValueError(...)` as it's easier to catch later if the code is used programmatically.